Add datadog-checks-downloader

trishankatdatadog opened the pull request:
### What does this PR do?

Add our secure `TUF` + `in-toto` downloader to the repository.

### Motivation

So that we no longer need to support our own fork of `pip`.

### Review checklist

- [ ] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [ ] Feature or bugfix has tests
- [ ] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)
- [ ] If PR adds a configuration option, it has been added to the configuration file.

### Additional Notes

Unit tests will come in the future.

trishankatdatadog commented:
We should be able to write a quite Python integration test that basically does the same job as `public-test.sh`.

gzussa commented:
@trishankatdatadog Please add proper labels

gzussa commented on integrations_core_downloader/README.md:
```diff
@@ -0,0 +1,3 @@
+# integrations-core-downloader
+
+A Datadog-specific secure downloader for integrations-core Python packages
```
Is this only for `core` (even long term)?

trishankatdatadog replied:
`core` only for now, but i will make it clear in doc

gzussa commented on integrations_core_downloader/integrations_core_downloader/data/integrations-core-repo/metadata/previous/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# https://steindom.com/articles/add-empty-directory-git-repository
```
Do we need this file?

trishankatdatadog replied:
yes, i have explained why in the file

gzussa commented on integrations_core_downloader/integrations_core_downloader/data/integrations-core-repo/targets/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# https://steindom.com/articles/add-empty-directory-git-repository
```
Do we need this file?

trishankatdatadog replied:
yes, i have explained why in the file

gzussa commented on integrations_core_downloader/README.md:
```diff
@@ -0,0 +1,3 @@
+# integrations-core-downloader
```
Let's detail how this is going to be used, in which version of the agent it will be used, how this package is different from all other packages in this repo, how it will be build as part of the agent and just why we need it and how it works (in toto, tuf, etc...). The more for our open source community friends, the better 😄

trishankatdatadog commented:
While I'm happy with the coverage report, I don't think it's accurate. It seems to be testing whether the tests are covered, and not necessarily all of the branches of the check itself.

trishankatdatadog commented:
@gzussa Addressed your comments, please let me know if you have further questions

@ofek We might need to make exceptions for `datadog_checks_downloader` in `ddev` and so on

trishankatdatadog commented on datadog_checks_downloader/datadog_checks/downloader/data/repo/metadata/current/.gitignore:
```diff
@@ -0,0 +1,5 @@
+# Ignore everything except the TUF root role metadata, and this .gitignore
+# file: https://jasonstitt.com/gitignore-whitelisting-patterns
+*
+root.json
```
Good catch!

trishankatdatadog commented on datadog_checks_downloader/datadog_checks/downloader/data/repo/metadata/previous/.gitignore:
```diff
@@ -0,0 +1,5 @@
+# This file is here, so that we can add this initially empty directory as part
```
We want to make these empty default directories when we install, so that we will download all files (TUF and in-toto metadata, as well as wheels) in there later, you see?

zippolyte replied:
I'm not sure I get it. Can't we create them on the fly ?

trishankatdatadog commented:
Ok, before we merge this, @ofek has asked that I replaced `click` with `argparse` in order to remove unnecessary dependencies. Let me work on that pronto...

trishankatdatadog commented:
@gzussa @zippolyte @ofek Done! 🚀 

AppVeyor randomly failed to connect to GitHub 🤦‍♂️

ofek commented on datadog_checks_downloader/.coveragerc:
```diff
@@ -0,0 +1,13 @@
+[run]
```
You can remove this file since we already have one

ofek commented on datadog_checks_downloader/tox.ini:
```diff
@@ -0,0 +1,27 @@
+minversion = 2.0
+skip_missing_interpreters = true
+envlist =
+    py{36,37}-datadog-checks-downloader
```
```suggestion
    py{27,36,37}
```

ofek commented on datadog_checks_downloader/tox.ini:
```diff
@@ -0,0 +1,27 @@
+deps =
+  -e../datadog_checks_base[deps]
+  -rrequirements-dev.txt
+  pytest-cov
```
```suggestion
```

ofek commented on datadog_checks_downloader/tox.ini:
```diff
@@ -0,0 +1,27 @@
+  pytest-cov
+commands =
+  pip install -r requirements.in
+  py.test --cov=datadog_checks tests/ --cov-fail-under=80 -s
```
```suggestion
  pytest -v
```

ofek commented on datadog_checks_downloader/pytest.ini:
```diff
@@ -0,0 +1,3 @@
+[pytest]
```
Let's remove?
